README
TwitShell

TwitShell is a microblogging client that runs via the command-line. It
supports many microblogging platform such as Twitter and Identi.ca.

The main purpose of TwitShell is to supply an simple tool to fetch
user's posted data (statuses, direct messages, etc...) and post data (status
update, direct messages) via a command-line interface.

TwitShell supports many microblogging features such as user timeline
and friends timeline, direct messages (both sent and received), status
update and direct messages dispatch.

TwitShell's code is based on the specifications of Twitter API.


INSTALLATION

To install TwitShell, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc TwitShell

or

    perldoc TwitCurses

You can also look for information at:

    TwitShell homepage
        http://alexlog.co.cc/projects/twitshell

    GitHub page
        http://github.com/AlexBio/TwitShell

    RT: CPAN's request tracker

        http://rt.cpan.org/NoAuth/Bugs.html?Dist=TwitShell

    AnnoCPAN: Annotated CPAN documentation

        http://annocpan.org/dist/TwitShell

    CPAN Ratings

        http://cpanratings.perl.org/d/TwitShell

    Search CPAN

        http://search.cpan.org/dist/TwitShell/

LICENSE AND COPYRIGHT

Copyright (C) 2010 Alessandro Ghedini

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.



Changes
commit a70f2707e9d76a55dc1edf79d375c49719f411cb
Author: Alessandro Ghedini <al3xbio@gmail.com>
Date:   Thu Jun 17 22:07:03 2010 +0200

    fix version conflict with PAUSE

commit 769298311181a07e2a8a7534bd6e5fa0b287a3d2
Author: Alessandro Ghedini <al3xbio@gmail.com>
Date:   Thu Jun 17 22:03:16 2010 +0200

    add use warnings

commit 661c00010b143885f12d69a0f6ae69ab2e0ebb7c
Author: Alessandro Ghedini <al3xbio@gmail.com>
Date:   Thu Jun 17 21:51:31 2010 +0200

    update README

commit b3c01bec28d1d7c62f5f6a8abb7153e8d027bcc7
Author: Alessandro Ghedini <al3xbio@gmail.com>
Date:   Thu Jun 17 21:49:37 2010 +0200

    update Changes

commit 62cefb0ec34e2abc31fe7290d1ff6fa4bfb4080f
Author: Alessandro Ghedini <al3xbio@gmail.com>
Date:   Thu Jun 17 21:49:12 2010 +0200

    new upstream release 4.07

commit 82f5ebbaa9a45ea50abfe0f609de63e83df7f85e
Author: Alessandro Ghedini <al3xbio@gmail.com>
Date:   Thu Jun 17 21:43:14 2010 +0200

    fix bug in which caused TwitCurses to not show first last post

commit fa7ce98b53b0efa0b9f113832b08bd8806e37aee
Author: Alessandro Ghedini <al3xbio@gmail.com>
Date:   Thu Jun 17 21:39:24 2010 +0200

    set friends timeline as default in TwitCurses

commit 478d75d0f932d079d019ff5091ea4d68624e2e2b
Author: Alessandro Ghedini <al3xbio@gmail.com>
Date:   Thu Jun 17 21:38:53 2010 +0200

    add proper use of Perl OOP

commit 6b52d9b82639c8378d356d1a2ba19e7a9bec4aaf
Author: Alessandro Ghedini <al3xbio@gmail.com>
Date:   Thu Jun 17 20:02:54 2010 +0200

    improve documentation style

commit bbbeff0d9acdddbff31e28cf97ea0408ea0e4fed
Author: Alessandro Ghedini <al3xbio@gmail.com>
Date:   Thu Jun 17 19:56:11 2010 +0200

    add proper use of bless

commit 6f1fc05909674feda775feb68a5141ba28da31d5
Author: Alessandro Ghedini <al3xbio@gmail.com>
Date:   Sun May 16 19:44:30 2010 +0200

    reupload Makefile.PL

commit 30cbbbb4bf4d7e654750cc85225d8583015bd932
Author: Alessandro Ghedini <al3xbio@gmail.com>
Date:   Sun May 16 19:42:54 2010 +0200

    update  MANIFEST and Makefile.PL

commit 56ddb3918b32a69efa3527101352ba48eb6ccd30
Author: Alessandro Ghedini <al3xbio@gmail.com>
Date:   Sun May 16 19:41:46 2010 +0200

    remove .pl extension and make scripts executable

commit eb4999ff8125ca8d11ee71a9082ed99a6ec42367
Author: Alessandro Ghedini <al3xbio@gmail.com>
Date:   Sun May 16 15:30:37 2010 +0200

    update Changes

commit f75c1fbeda181142023d392f4a1a09d2299362a6
Author: Alessandro Ghedini <al3xbio@gmail.com>
Date:   Sun May 16 15:30:11 2010 +0200

    add tests

commit 7ab6e61d004d14f6fa694dd5299f933c82227a38
Author: Alessandro Ghedini <al3xbio@gmail.com>
Date:   Sun May 16 14:51:07 2010 +0200

    add use strict + bug fix

commit fdd4c9d0b56754130f3714184dbb8eff2daeaa41
Author: Alessandro Ghedini <al3xbio@gmail.com>
Date:   Sun May 16 14:48:45 2010 +0200

    fix wrong password set bug

commit dc791c28353fd74acf0ecb03f2ecd8a8939b9957
Author: Alessandro Ghedini <al3xbio@gmail.com>
Date:   Sun May 16 14:42:37 2010 +0200

    make CLASS -> get_pwd() to directly set CLASS -> {'password'}

commit 2c1ff29f33605a28e1bc128ec9e19ba37e439e77
Author: Alessandro Ghedini <al3xbio@gmail.com>
Date:   Sun May 16 14:39:16 2010 +0200

    add documentation to modules

commit dac164538d6a0a45e74062726253f9034b7aa008
Author: Alessandro Ghedini <al3xbio@gmail.com>
Date:   Sat May 15 21:01:34 2010 +0200

    add Changes MANIFEST README

commit 82588f7af098ed867afa05b64acffa22af368a35
Author: Alessandro Ghedini <al3xbio@gmail.com>
Date:   Sat May 15 20:50:44 2010 +0200

    add POD documentation

commit 722907914bd22c7f36dfc38eae3ac1e020e76367
Author: Alessandro Ghedini <al3xbio@gmail.com>
Date:   Sat May 15 18:45:09 2010 +0200

    rename script to scripts

commit 24d49ee25419afd6c14b39536a65619374c0bcea
Author: Alessandro Ghedini <al3xbio@gmail.com>
Date:   Sat May 15 18:43:08 2010 +0200

    add TwitCurses

commit 7e2eca8bb5ffea682a2fc8222c6630e7feb839d4
Author: Alessandro Ghedini <al3xbio@gmail.com>
Date:   Sat May 15 16:48:13 2010 +0200

    add verify_credentials()

commit 4a0812ffa5f212629b3b950a789dff1e618f3230
Author: Alessandro Ghedini <al3xbio@gmail.com>
Date:   Sat May 15 16:30:58 2010 +0200

    fix undef user'name and screen_name

commit 5ad3361c3568de175d3ad35f7bbd70532d9a07d4
Author: Alessandro Ghedini <al3xbio@gmail.com>
Date:   Sat May 15 16:19:23 2010 +0200

    add App::TwitShell::Resources::Shorten

commit 4a4bfd4a6794d08e96c94aced49c135f9f396322
Author: Alessandro Ghedini <al3xbio@gmail.com>
Date:   Sat May 15 15:45:26 2010 +0200

    add url shortening using U.NU

commit 90cafc5b6b9c034fba9738feeb7d533fe3c8e92e
Author: Alessandro Ghedini <al3xbio@gmail.com>
Date:   Sat May 15 15:18:41 2010 +0200

    add relative_time() to App::TwitShell::Resources

commit 7629e1d446777c8a9c767cbf13ddefc0c6c84aee
Author: Alessandro Ghedini <al3xbio@gmail.com>
Date:   Sat May 15 15:11:45 2010 +0200

    add format output in TwitShell

commit ebe957250a9b465bce2569a3770b23333a71e3cb
Author: Alessandro Ghedini <al3xbio@gmail.com>
Date:   Sat May 15 14:46:52 2010 +0200

    add better hash parsing + error handling

commit c17fcb66a1b85fc8e9b579740eb408ce7771fc76
Author: Alessandro Ghedini <al3xbio@gmail.com>
Date:   Fri May 14 22:09:58 2010 +0200

    add use of format_post() in App::TwitShell::Resources::Send

commit 0adad9c76e3b9037c716514803920c8f23cd8d23
Author: Alessandro Ghedini <al3xbio@gmail.com>
Date:   Fri May 14 22:09:03 2010 +0200

    move format_post() to App::TwitShell::Resources

commit c2de9cfdf29ea11226d97e9e1c5a988b596cb0fa
Author: Alessandro Ghedini <al3xbio@gmail.com>
Date:   Fri May 14 22:07:15 2010 +0200

    add basic non-empty check on user input

commit 8055222b2bc6f25caedc5af287f3bbb90c0765ea
Author: Alessandro Ghedini <al3xbio@gmail.com>
Date:   Fri May 14 21:59:07 2010 +0200

    add format_post()

commit d8d3618b14b88ef25b58c10dbf755cd3ef5aa609
Author: Alessandro Ghedini <al3xbio@gmail.com>
Date:   Fri May 14 21:45:56 2010 +0200

    add minor changes


Makefile.PL
use strict;
use warnings;
use ExtUtils::MakeMaker;
 
WriteMakefile(
	'NAME'		=> 'TwitShell',
	'AUTHOR' 	=> 'Alessandro Ghedini <al3xbio@gmail.com>',
	'ABSTRACT' 	=> 'Command-line microblogging client.',
	'LICENSE' 	=> 'GPLv3',
	'VERSION_FROM'	=> 'scripts/TwitShell',

	PM            	=> {
		'lib/App/TwitShell.pm'        		 => 'blib/lib/App/TwitShell.pm',
		'lib/App/TwitShell/Resources.pm'         => 'blib/lib/App/TwitShell/Resources.pm',
		'lib/App/TwitShell/Resources/Get.pm'     => 'blib/lib/App/TwitShell/Resources/Get.pm',
		'lib/App/TwitShell/Resources/Send.pm'    => 'blib/lib/App/TwitShell/Resources/Send.pm',
		'lib/App/TwitShell/Resources/Shorten.pm' => 'blib/lib/App/TwitShell/Resources/Shorten.pm',
	},

	EXE_FILES	=> [
		'scripts/TwitShell',
		'scripts/TwitCurses'
	],
 
	'MAN1PODS' 	=> {
		'scripts/TwitShell'  => 'blib/man1/TwitShell.1',
		'scripts/TwitCurses' => 'blib/man1/TwitCurses.1'
	},
	
	'PREREQ_PM'	=> {
		'HTTP::Request::Common'  => 0,
		'LWP::UserAgent' 	 => 0,
		'JSON' 			 => 0,
		'Time::Out' 		 => 0,
		'Curses' 		 => 0,
		'DateTime::Format::HTTP' => 0,
		'Getopt::Long'		 => 0
	},

);


lib/App/TwitShell.pm
package App::TwitShell;

use App::TwitShell::Resources;
use App::TwitShell::Resources::Get;
use App::TwitShell::Resources::Send;

use Getopt::Long;

use strict;
use warnings;

=head1 NAME

App::TwitShell - Container for TwitShell objects's functions.

=head1 VERSION

Version 4.08

=cut

our $VERSION = 4.08;

=head1 SYNOPSIS

App::TwitShell contains some common functions used with the TwitShell
client object.

=head1 METHODS

=head2 new()

Create an instance of the Twitter client and assign some default values.

=cut

sub new {
	my ($class) = @_;
	
	my $self = bless({configfile => $ENV{'HOME'}.'/.twitshellrc',
			  count	     => 20,
			  network    => 'twitter',
			  shorten    => 0,
			  refresh    => 0,
			  secs       => 45
			 }, $class);
	return $self;
}

=head2 parse_config()

Parse the .twitshellrc file.

=cut

sub parse_config {
	my $self = shift;

	open(CONFIG, $self -> {'configfile'}) or error ("$!");
	while (<CONFIG>) {
		chomp;                  # no newline
		s/#.*//;                # no comments
		s/^\s+//;               # no leading white
		s/\s+$//;               # no trailing white
		next unless length;     # anything left?
		my ($var, $value) = split(/\s*=\s*/, $_, 2);
		$self -> {$var} = $value;
	}
}

=head2 parse_arguments()

Parse the command-line arguments.

=cut

sub parse_arguments {
	my $self = shift;
	
	my $getopt  = Getopt::Long::GetOptions(
			'action=s'   => \$self -> {'action'},
			'id=i' 	     => \$self -> {'id'},
			'msg=s'      => \$self -> {'msg'},
			'username=s' => \$self -> {'username'},
			'network=s'  => \$self -> {'network'},
			'count=i'    => \$self -> {'count'},
			'refresh=i'  => \$self -> {'refresh'},
			'secs=i'     => \$self -> {'username'},
			'shorten=i'  => \$self -> {'username'},
			'recipient=s'=> \$self -> {'recipient'}
	);
}

=head2 run()

Execute the action contant in 'action'.

=cut

sub run {
	my $self = shift;

	error("Missing 'username'") if $self -> {'username'} eq '';
	error("Missing 'network'")  if $self -> {'network'}  eq '';
	
	if ($self -> {'action'} eq 'user') {
		$self -> {'posts'} = get_user_posts($self);
							       
	} elsif ($self -> {'action'} eq 'friends') {
		$self -> {'posts'} = get_friends_posts($self);

	} elsif ($self -> {'action'} eq 'inbox') {
		$self -> {'posts'} = get_inbox($self);

	} elsif ($self -> {'action'} eq 'outbox') {
		$self -> {'posts'} = get_outbox($self);

	} elsif ($self -> {'action'} eq 'single') {
		error("Missing 'id'") if $self -> {'id'} eq '';
		
		$self -> {'posts'} = get_single_post($self);

	} elsif ($self -> {'action'} eq 'update') {
		error("Missing 'msg'") if $self -> {'msg'} eq '';
		
		$self -> {'posts'} = update($self);

	} elsif ($self -> {'action'} eq 'send') {
		error("Missing 'msg'")       if $self -> {'msg'}       eq '';
		error("Missing 'recipient'") if $self -> {'recipient'} eq '';
		
		$self -> {'posts'} = send_dm($self);
	}
}

=head2 verify_credentials()

Verify the user's credentials via the web APIs, using 'username'
and 'password'.

=cut

sub verify_credentials {
	my $self = shift;

	my $api = get_api($self -> {'network'});

	my $url 	= $api -> {'url'}."/account/verify_credentials.json";
	my $response 	= auth_get_request($url, $self -> {'username'}, $self -> {'password'}, $api);
	
	my $json_text 	= parse_response($response);

	my $error 	= $json_text -> {error};

	error("$error") if ($error ne '');
}

=head2 get_pwd()

Request user's password.

=cut

sub get_pwd {
	my $self = shift;
	
	print "Password: ";
	system('stty','-echo');
	chop(my $pwd=<STDIN>);
	system('stty','echo');
	print "\n";
	
	$self -> {'password'} = $pwd;
}

=head2 error( $msg )

Print formatted errors and die

=cut

sub error {
	my $msg = shift;

	die "ERROR: $msg\n";
}

=head2 get_data()

Return content of a TwitShell object

=cut

sub get_data {
	my $self = shift;

	my %data = %$self;

	return \%data;
}

=head1 AUTHOR

Alessandro Ghedini, C<< <alexbio at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-twitshell at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=TwitShell>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc App::TwitShell

You can also look for information at:

=over 4

=item * TwitShell homepage

L<http://alexlog.co.cc/projects/twitshell>

=item * GitHub page

L<http://github.com/AlexBio/TwitShell>

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=TwitShell>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/TwitShell>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/TwitShell>

=item * Search CPAN

L<http://search.cpan.org/dist/TwitShell/>

=back

=head1 LICENSE AND COPYRIGHT

Copyright 2010 Alessandro Ghedini.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.


=cut

1; # End of App::TwitShell


lib/App/TwitShell/Resources/Get.pm
package App::TwitShell::Resources::Get;

use App::TwitShell::Resources;

use strict;
use warnings;
use base 'Exporter';

our @EXPORT = qw(get_single_post get_user_posts get_friends_posts
		 get_inbox get_outbox);

=head1 NAME

App::TwitShell::Resources::Get - Container for data retrieving function (API
interface).

=head1 VERSION

Version 4.08

=cut

our $VERSION = 4.08;

=head1 SYNOPSIS

App::TwitShell::Resources::Get is the Twitter's API interface for data
retrieving.

=head1 FUNCTIONS

=head2 get_single_post( $id, $network )

Retrieve a single post.

=cut

sub get_single_post {
	my $client = shift;
	my $data = $client -> get_data;

	my $api = get_api($data -> {'network'});
	
	my $url = $api -> {'url'}."/statuses/show/".$data -> {'id'}.".json";

	my $response  = get_request($url);
	my $json_text = parse_response($response);
	
	my $posts = parse_post_hash($json_text, 'status');
	
	return $posts;
}

=head2 get_user_posts( $id, $screen_name, $count, $network )

Retrieve given user's post (user_timeline).

=cut

sub get_user_posts {
	my $client = shift;
	my $data = $client -> get_data;

	my $api = get_api($data -> {'network'});

	my $options = "screen_name=".$data -> {'screen_name'}."&count=".$data -> {'count'};
	my $url = $api -> {'url'}."/statuses/user_timeline.json";
	
	my $response  = get_request("$url?$options");
	my $json_text = parse_response($response);
	
	my $posts = parse_post_hash($json_text, 'status');
	
	return $posts;
}

=head2 get_friends_posts( $username, $password, $count, $network )

Retrieve given user's friends post (friends_timeline). Require authentication.

=cut

sub get_friends_posts {
	my $client = shift;
	my $data = $client -> get_data;

	my $api = get_api($data -> {'network'});

	my $url = $api -> {'url'}."/statuses/friends_timeline.json?count=".$data -> {'count'};

	my $response  = auth_get_request($url, $data -> {'username'},
					 $data -> {'password'}, $api);

	my $json_text = parse_response($response);
	
	my $posts = parse_post_hash($json_text, 'status');
	
	return $posts;
}

=head2 get_inbox( $username, $password, $count, $network )

Retrieve given user's received direct messages (direct_messages). Require authentication.

=cut

sub get_inbox {
	my $client = shift;
	my $data = $client -> get_data;

	my $api = get_api($data -> {'network'});

	my $url = $api -> {'url'}."/direct_messages.json?count=".$data -> {'count'};
	
	my $response  = auth_get_request($url, $data -> {'username'},
					 $data -> {'password'}, $api);

	my $json_text = parse_response($response);
	
	my $posts = parse_post_hash($json_text, 'dm');
	
	return $posts;
}

=head2 get_outbox( $username, $password, $count, $network )

Retrieve given user's sent direct messages (direct_messages/sent).
Require authentication.

=cut

sub get_outbox {
	my $client = shift;
	my $data = $client -> get_data;

	my $api = get_api($data -> {'network'});
	
	my $url = $api -> {'url'}."/direct_messages/sent.json?count=".$data -> {'count'};
	
	my $response  = auth_get_request($url, $data -> {'username'},
					 $data -> {'password'}, $api);

	my $json_text = parse_response($response);
	
	my $posts = parse_post_hash($json_text, 'dm');
	
	return $posts;
}

=head1 AUTHOR

Alessandro Ghedini, C<< <alexbio at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-twitshell at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=TwitShell>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc App::TwitShell::Resources::Get

You can also look for information at:

=over 4

=item * TwitShell homepage

L<http://alexlog.co.cc/projects/twitshell>

=item * GitHub page

L<http://github.com/AlexBio/TwitShell>

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=TwitShell>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/TwitShell>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/TwitShell>

=item * Search CPAN

L<http://search.cpan.org/dist/TwitShell/>

=back

=head1 LICENSE AND COPYRIGHT

Copyright 2010 Alessandro Ghedini.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.


=cut

1; # End of App::TwitShell::Resources::Get



lib/App/TwitShell/Resources/Send.pm
package App::TwitShell::Resources::Send;

use App::TwitShell::Resources;
use App::TwitShell::Resources::Shorten;

use strict;
use warnings;
use base 'Exporter';

our @EXPORT = qw(update send_dm);

=head1 NAME

App::TwitShell::Resources::Send - Container for data posting function (API
interface).

=head1 VERSION

Version 4.08

=cut

our $VERSION = 4.08;

=head1 SYNOPSIS

App::TwitShell::Resources::Send is the Twitter's API interface for data
posting.

=head1 FUNCTIONS

=head2 update( $client )

Update authenticating user's status.

=cut

sub update {
	my $client = shift;
	my $data = $client -> get_data;

	my $api = get_api($data -> {'network'});

	my $msg = auto_shorten($data -> {'msg'}) if ($data -> {'shorten'} == 1);

	my $url = $api -> {'url'}."/statuses/update.json";

	my %request = ('status', $msg);
	my $response  = auth_post_request($url, $data -> {'username'},
					  $data -> {'password'}, \%request, $api);
	
	my $json_text = parse_response($response);
	
	my $posts = parse_post_hash($json_text, 'status');
	
	return $posts;
}

=head2 send_dm( $username, $password, $msg, $recipient, $shorten, $network )

Send a direct message from authenticating user.

=cut

sub send_dm {
	my $client = shift;
	my $data = $client -> get_data;

	my $api = get_api($data -> {'network'});

	my $msg = auto_shorten($data -> {'msg'}) if ($data -> {'shorten'} == 1);

	my $url = $api -> {'url'}."/direct_messages/new.json?screen_name=".$data -> {'recipient'};

	my %request = ('text', $msg);
	my $response  = auth_post_request($url, $data -> {'username'},
					  $data -> {'password'}, \%request, $api);
	
	my $json_text = parse_response($response);
	
	my $posts = parse_post_hash($json_text, 'dm');
	
	return $posts;
}

=head2 auto_shorten( $text )

Parse messages, extract URLs adn shorten them.

=cut	

sub auto_shorten() {
	my $text = shift;
	
	$text .= " "; # unelegant, but works well
	while ($text =~ m/(http\:\/\/.*?) /g) {
		if (index($1, "http://u.nu") == -1) {
			my $short = App::TwitShell::Resources::Shorten -> shorten($1, 'u.nu');
			$text =~ s/$1/$short/g;
		}
	}

	# remove trailing whitespace
	$/ = " ";
	chomp($text);

	return $text;
}

=head1 AUTHOR

Alessandro Ghedini, C<< <alexbio at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-twitshell at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=TwitShell>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc App::TwitShell::Resources::Send

You can also look for information at:

=over 4

=item * TwitShell homepage

L<http://alexlog.co.cc/projects/twitshell>

=item * GitHub page

L<http://github.com/AlexBio/TwitShell>

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=TwitShell>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/TwitShell>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/TwitShell>

=item * Search CPAN

L<http://search.cpan.org/dist/TwitShell/>

=back

=head1 LICENSE AND COPYRIGHT

Copyright 2010 Alessandro Ghedini.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.


=cut

1; # End of App::TwitShell::Resources::Send


lib/App/TwitShell/Resources/Shorten.pm
package App::TwitShell::Resources::Shorten;

use App::TwitShell::Resources;

use strict;
use warnings;
use base 'Exporter';

our @EXPORT = qw(shorten);

=head1 NAME

App::TwitShell::Resources::Shorten - Container for URL shortening functions.

=head1 VERSION

Version 4.08

=cut

our $VERSION = 4.08;

=head1 SYNOPSIS

App::TwitShell::Resources::Shorten is the Url shortening services' API
interface.

=head1 FUNCTIONS

=head2 shorten( $longurl, $service )

Shorten longurl using selected web service.

=cut

sub shorten {
	my ($self, $longurl, $service) = @_;

	my %services = ('u.nu' => "http://u.nu/unu-api-simple?url=$longurl");

	my $url = $services{$service};

	my @data = split('\n\n', App::TwitShell::Resources -> get_request($url));
	chomp($data[1]);

	return $data[1];
}

=head1 AUTHOR

Alessandro Ghedini, C<< <alexbio at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-twitshell at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=TwitShell>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc App::TwitShell::Resources::Shorten

You can also look for information at:

=over 4

=item * TwitShell homepage

L<http://alexlog.co.cc/projects/twitshell>

=item * GitHub page

L<http://github.com/AlexBio/TwitShell>

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=TwitShell>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/TwitShell>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/TwitShell>

=item * Search CPAN

L<http://search.cpan.org/dist/TwitShell/>

=back

=head1 LICENSE AND COPYRIGHT

Copyright 2010 Alessandro Ghedini.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.


=cut

1; # End of App::TwitShell::Resources::Shorten



lib/App/TwitShell/Resources.pm
package App::TwitShell::Resources;

use HTTP::Request::Common;
use LWP::UserAgent;
use JSON -support_by_pp;
use DateTime::Format::HTTP;

use strict;
use warnings;
use base 'Exporter';

our @EXPORT = qw(get_api parse_response parse_post_hash 
		 get_request auth_get_request auth_post_request);

=head1 NAME

App::TwitShell::Resources - Container for resources retrieving and parsing
functions.

=head1 VERSION

Version 4.08

=cut

our $VERSION = 4.08;

=head1 SYNOPSIS

App::TwitShell::Resources contains every common function for HTTP
request and data parsing/formatting.

=head1 FUNCTIONS

=head2 get_request( $url )

Execute a simple GET request.

=cut

sub get_request {
	my $url = shift;
	my $ua = LWP::UserAgent -> new;

	my $response = $ua -> request(GET $url) -> as_string;

	return $response;
}

=head2 auth_get_request( $url, $usr, $pwd, $api )

Execute a GET request with HTTP authorization.

=cut

sub auth_get_request {
	my ($url, $usr, $pwd, $api) = @_;
	my $ua = LWP::UserAgent -> new;
	
	$ua -> credentials($api -> {'endpoint'}, $api -> {'realm'}, $usr, $pwd);
	my $response = $ua -> request(GET $url) -> as_string;

	return $response;
}

=head2 auth_post_request( $url, $usr, $pwd, \%request, $api )

Execute a POST request with HTTP authorization.

=cut

sub auth_post_request {
	my ($url, $usr, $pwd, $request, $api) = @_;
	my $ua = LWP::UserAgent->new;

	$ua -> credentials($api -> {'endpoint'}, $api -> {'realm'}, $usr, $pwd);
	my $response = $ua -> request(POST $url, [%$request]) -> as_string;

	return $response;
}

=head2 parse_response( $data )

Parse a HTTP response splitting JSON data, and return a JSON object.

=cut

sub parse_response {
	my @data = split('\n\n', shift);
	
	my $json = new JSON;
	my $json_text = $json -> decode($data[1]);

	if (ref $json_text ne "ARRAY") {
		die("ERROR: ".$json_text -> {'error'}."\n") if $json_text -> {'error'};
	}
	
	return $json_text;
}

=head2 parse_post_hash( $data, $type )

Parse a JSON object, and return an array of posts.

=cut

sub parse_post_hash {
	my ($data, $type) = @_;

	my @posts;
	
	if ($data =~ /array/i) {
		foreach my $status(@{$data}) {
			my $post = format_post($status, $type);
			push(@posts, $post);	
		}
	} else {
		my $post = format_post($data, $type);
		push(@posts, $post);
	}

	return \@posts;
}

=head2 get_api( $network )

Return an API hash based on selected network.

=cut

sub get_api {
	my $network = shift;

	my %api;
	
	if ($network eq "twitter") {
		$api{'endpoint'} = "api.twitter.com:80";
		$api{'path'}     = "1";
		$api{'realm'}    = "Twitter API";
	} elsif ($network eq "identi.ca") {
		$api{'endpoint'} = "identi.ca:80";
		$api{'path'}     = "api";
		$api{'realm'}    = "Identi.ca API";
	}

	$api{'url'} = "http://".$api{'endpoint'}."/".$api{'path'};

	return \%api;
}

=head2 format_post( $status, $type )

Format a single post, from a hash, based on selected post type.

=cut

sub format_post {
	my ($status, $type) = @_;

	my %out;

	if ($status -> {'error'}) {
		$out{'err'} = 1;
		$out{'err_msg'} = $status -> {error};
	} else {
		$out{'err'} = 0;
		
		$out{'text'} 	 = $status -> {text};
		$out{'date'} 	 = $status -> {created_at};
		$out{'relative'} = relative_time($status -> {created_at});
		$out{'id'}   	 = $status -> {id};

		if ($type eq 'status') {
			$out{'user'} 	     = $status -> {user} -> {name};
			$out{'screen_name'}  = $status -> {user} -> {screen_name};
		} elsif ($type eq 'dm') {
			$out{'user'} 	     = $status -> {sender} -> {name};
			$out{'screen_name'}  = $status -> {sender} -> {screen_name};
		}
	}

	return \%out;
}

=head2 relative_time( $time_value )

Return the relative time from a standard data.

=cut

sub relative_time {
	my $time_value = shift;

	my $relative_to = DateTime -> now() -> epoch;
	
	$time_value =~ s/\+0000/GMT/;
	$time_value = DateTime::Format::HTTP -> parse_datetime($time_value) -> epoch;
	
	$relative_to = $relative_to -> epoch if ref($relative_to) =~ /DateTime/;
	my $delta = $relative_to - $time_value;
	
	if($delta < 60) {
		return 'less than a minute ago';
	} elsif($delta < 120) {
		return 'about a minute ago';
	} elsif($delta < (45*60)) {
		return int($delta / 60) . ' minutes ago';
	} elsif($delta < (90*60)) {
		return 'about an hour ago';
	} elsif($delta < (24*60*60)) {
		return 'about ' . int($delta / 3600) . ' hours ago';
	} elsif($delta < (48*60*60)) {
		return '1 day ago';
	} else {
		return int($delta / 86400) . ' days ago';
	}
}

=head1 AUTHOR

Alessandro Ghedini, C<< <alexbio at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-twitshell at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=TwitShell>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc App::TwitShell::Resources

You can also look for information at:

=over 4

=item * TwitShell homepage

L<http://alexlog.co.cc/projects/twitshell>

=item * GitHub page

L<http://github.com/AlexBio/TwitShell>

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=TwitShell>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/TwitShell>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/TwitShell>

=item * Search CPAN

L<http://search.cpan.org/dist/TwitShell/>

=back

=head1 LICENSE AND COPYRIGHT

Copyright 2010 Alessandro Ghedini.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.


=cut

1; # End of App::TwitShell::Resources
